src/Host/Client/Impl/Host/RHostConnector.cs

```diff
@@ -28,38 +30,68 @@ internal abstract class RHostConnector : IRHostConnector {
 
         private readonly LinesLog _log;
-        private HttpClient _broker;
         private string _interpreterId;
 
-        protected HttpClient Broker => _broker;
+        protected HttpClientHandler HttpClientHandler { get; private set; }
+
+        protected HttpClient HttpClient { get; private set; }
 
         protected RHostConnector(string interpreterId) {
             _interpreterId = interpreterId;
             _log = new LinesLog(FileLogWriter.InTempFolder("Microsoft.R.Host.BrokerConnector"));
         }
 
-        protected void CreateHttpClient() {
-            _broker = new HttpClient(GetHttpClientHandler()) {
+        protected void CreateHttpClient(Uri baseAddress, ICredentials credentials) {
+            HttpClientHandler = new HttpClientHandler {
+                PreAuthenticate = true,
+                Credentials = credentials
+            };
+
+            HttpClient = new HttpClient(HttpClientHandler) {
+                BaseAddress = baseAddress,
                 Timeout = TimeSpan.FromSeconds(30)
             };
 
-            _broker.DefaultRequestHeaders.Accept.Clear();
-            _broker.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
+            HttpClient.DefaultRequestHeaders.Accept.Clear();
+            HttpClient.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
         }
 
-        protected abstract HttpClientHandler GetHttpClientHandler();
-
-        protected abstract void ConfigureWebSocketRequest(HttpWebRequest request);
-
         public void Dispose() {
             DisposableBag.TryMarkDisposed();
         }
 
+        /// <summary>
+        /// Called before issuing an authenticated HTTP request. Implementation can refresh <see cref="HttpClientHandler.Credentials"/> if necessary.
+        /// </summary>
+        /// <remarks>
+        /// For every call to this method, there will be a follow-up call to either <see cref="OnAuthenticationSucceeded"/> 
+        /// or to <see cref="OnAuthenticationFailed"/> to indicate the result of authentication.
+        /// </remarks>
+        /// <exception cref="OperationCanceledException">
+        /// Retrieval of credentials was canceled by the user (for example, by clicking the "Cancel" button in the dialog).
+        /// Usually, this indicates that the operation that asked for credentials should be canceled as well.
+        /// </exception>
+        protected abstract void UpdateCredentials();
+
+        /// <summary>
+        /// Called whenever HTTP authentication succeeds.
+        /// </summary>
+        protected abstract void OnAuthenticationSucceeded();
+
+        /// <summary>
+        /// Called whenever HTTP authentication fails (i.e. server returns HTTP 403).
+        /// </summary>
+        /// <returns>
+        /// <see langword="true"/> if <see cref="Credentials"/> were updated, and HTTP request should be retried.
```

Comment:
The comment is misleading now, since `OnAuthenticationFailed` doesn't actually update credentials. I'll replace this with a simpler arrangement and adjust the comment.